### PR TITLE
Orchestrator: write registry entries resulting from overrides to their own versioned path

### LINKED ISSUE #30669: Bug: Update Orchestrator to write registry entry according to its own version
## Problem
If the metadata (v.2.0.0) has a dockerImageTag override for cloud (v1.0.0) then if you look up registry entry v2.0.0 for cloud you will actually get v1.0.0

## solution
when writing a registry entry from a metadata file. Use the registry entry version in the write path not the path of the metadata file.

### PR DESCRIPTION
**Pull request by bnchrch**
## Problem
Currently a state exists where you can request the registry entry for source-s3/v1.0.0/cloud.json and recieve a payload with the docker image source-s3/v0.0.1

closes #30669
## Solution
Write the resulting registry entry to a path based on its version not the metadatas version

### CODE REVIEW

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/hacks.py [hunk 1]
```diff
@@ -44,48 +31,56 @@ def _check_for_invalid_write_path(write_path: str):
...
+        Example:
+        If metadata file for source-posgres is at version 2.0.0 but there is a override for the cloud registry
+        that changes the docker repository to source-postgres-strict-encrypt and the version to 1.0.0
+
+        Then we will have a metadata file written to:
+        gs://my-bucket/metadata/source-postgres/2.0.0/metadata.yaml
+
+        and registry entries written to:
+        gs://my-bucket/metadata/source-postgres/2.0.0/oss.json
+        gs://my-bucket/metadata/source-postgres-strict-encrypt/1.0.0/cloud.json
```
Comment (pedroslopez) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/hacks.py:65]: Why do we need to rewrite the registry entry for the old version? Should we _only_ modify `latest`?

My thought here is that if when we published 1.0.0 there were some different fields like suggested streams applicable for 1.0.0, now we would be grabbing the suggested streams from 2.0.0 and using those instead - is that right?

My feeling is that we should leave the old one be and only tweak latest when dealing with overrides, but curious to know what your thoughts are here and if there's some benefit I'm missing.

**Review by pedroslopez — COMMENTED** [review]
Didn't dive into the code quite yet, just a question on the overall approach

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/hacks.py (hunk 1 above)
Comment (pedroslopez) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/hacks.py:65]: I think if i had a magic wand the ideal world-view would be to separate this override stuff from rollbacks/version changes. Reason being that if I want to now use 1.0.0, I really do want all the versioned metadata that was present in 1.0.0 to be used, not just some pieces of it (e.g. spec, image tag).

Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/hacks.py:65]: @pedroslopez Good news is I think you, me and Ella all agree on that ideal world view.

So heres the contextual facts from Ella and I's chat last week :
1. We only have overrides today
2. That leads to all configuration settings that affect our Registries live in the master branch
3. This is useful if you consider a case that the platform requires a new metadata field, we dont have to do any special backfilling of old registry entries / metadata files for overrode docker images
4. However this is not useful (specifically for rollbacks) as you have to be aware of if an overrode docker image had any metadata requirements 
5. and to add to this, Launch Darkly Overides behave differently to Registry Overrides. With launch darkly you can time travel as we reference the versioned registry entry directly.
6. LD refs combined with Metadata overrides (in the current state) lead to a hole where a LD override cannot find a Registry Entry for a strict-encrypt version.

What this leads us to believe is:
1. We need BOTH the ability to OVERRIDE metadata per registry and the ability to REF a different version of the metadata for the registry.
2. If refs become the prefered method then LD and Registry overrides dont deviate in behaviour which is a good thing
2. But! If we do introduce REFs this will require us to think hard about required registry fields, out of date refs (and how to update them), and global vs versioned metadata (tech spec territory)
3. And if were in tech spec territory then we also have the ability to revisit "Should we be using docker image information as our source of truth for versioning and paths"

**TLDR: We should do refs, but we need to explicitly design it, in the meantime we should plug this hole in our paths that LD references**

Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/hacks.py:65]: Good question @pedroslopez :

So the main culprit here is a few facts
1. Connectors are a combination of artifacts (dockerfiles) and metadata
1. We use docker image tags as connector versions
1. Metadata can change without a new docker image tag being required
1. There are two instances of a connector (cloud and oss) per version

This means that anytime we update the metadata for a connector, we need to apply that new metadata to all "latest" versions for cloud and oss.

For example
1. if the metadata for source-s3 is set so that oss get v2 and cloud gets v1
2. then platform says having field `xyz` is now mandatory
3. we need to update the metadata to include xyz so that the registry entry for v2 and the registry entry for v1 have xyz

Happy to jump on a call if that helps 😅

Comment (pedroslopez) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/hacks.py:65]: > 3. we need to update the metadata to include xyz so that the registry entry for v2 and the registry entry for v1 have xyz

@bnchrch I think this is the bit I don't quite understand - why does v1 now need xyz if we're not copying all of v1 metadata to latest when an override is in place but rather just tweaking the docker_image_tag field of whatever's latest?

Feels like we're in between two states:
1. The current world where we have a set of metadata and just override the docker_image_tag of that metadata
2. The ideal world where saying we want to use v1 means we use v1's metadata specific to that version

If we're in world view 1 and the definition of specifying an override is "use this other value for this field instead", then I don't think we should be doing anything with the existing data of other versions, but just make latest match with whatever override you set on these specific fields.

Does that make sense? Also happy to jump on a call

**Review by pedroslopez — APPROVED** [review]
